Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp

```diff
@@ -395,82 +397,114 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+
 #if ENABLE(ARKIT_INLINE_PREVIEW)
-    if (!configuration->modelElementCacheDirectory().isEmpty())
-        resolvedConfiguration->setModelElementCacheDirectory(resolveAndCreateReadWriteDirectoryForSandboxExtension(configuration->modelElementCacheDirectory()));
+    if (!directories.modelElementCacheDirectory.isEmpty())
+        directories.modelElementCacheDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(directories.modelElementCacheDirectory);
 #endif
-    if (!configuration->searchFieldHistoryDirectory().isEmpty())
-        resolvedConfiguration->setSearchFieldHistoryDirectory(resolveAndCreateReadWriteDirectoryForSandboxExtension(configuration->searchFieldHistoryDirectory()));
 
-    // Resolve file paths.
-    if (!configuration->cookieStorageFile().isEmpty()) {
-        auto resolvedCookieDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(FileSystem::parentPath(configuration->cookieStorageFile()));
-        resolvedConfiguration->setCookieStorageFile(FileSystem::pathByAppendingComponent(resolvedCookieDirectory, FileSystem::pathFileName(configuration->cookieStorageFile())));
+    if (!directories.cookieStorageFile.isEmpty()) {
+        auto resolvedCookieDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(FileSystem::parentPath(directories.cookieStorageFile));
+        directories.cookieStorageFile = FileSystem::pathByAppendingComponent(resolvedCookieDirectory, FileSystem::pathFileName(directories.cookieStorageFile));
     }
+}
+
+const WebsiteDataStoreConfiguration::Directories& WebsiteDataStore::resolvedDirectories()
+{
+    Locker resolveLocker { m_resolveLock };
+    return *m_resolvedDirectories;
+}
+
+void WebsiteDataStore::dispatchResolveDirectories()
+{
+    ASSERT(RunLoop::isMain());
 
-    // Default paths of WebsiteDataStore created with identifier are not under caches or tmp directory,
-    // so we need to explicitly exclude them from backup.
-    if (configuration->identifier()) {
-        Vector<String> allCacheDirectories = {
-            resolvedMediaCacheDirectory()
-            , resolvedNetworkCacheDirectory()
+    m_hasDispatchedResolveDirectories = true;
+    Ref resolveDirectoriesQueue = WorkQueue::create("com.apple.WebKit.WebsiteDataStore.resolveDirectories", WorkQueue::QOS::UserInteractive);
+    resolveDirectoriesQueue->dispatch([this, protectedThis = Ref { *this }, directories = m_configuration->directories().isolatedCopy()]() mutable {
+        resolveDirectories(directories);
+        {
+            Locker resolveLocker { m_resolveLock };
+            m_resolvedDirectories = crossThreadCopy(WTFMove(directories));
+            m_resolveDirectoriesCondition.notifyOne();
+        }
+        didResolveDirectories();
+        RunLoop::main().dispatch([protectedThis = WTFMove(protectedThis)] { });
+    });
+}
+
+void WebsiteDataStore::didResolveDirectories()
+{
+    Vector<String> allCacheDirectories;
+    auto directories = resolvedDirectories();
```

Comment:
Can we use `auto&` here? Note that you're off the main thread now so copying the struct and thus the Strings it holds is likely not safe.